File: clojure/P23.clj
; P23 Extract a given number of randomly selected elements from a list.

; Helper function
(defn remove-at [n l]
  (let [[a [x & b]] (split-at n l)]
    [(concat a b) x]))

; Building a target list while removing items from the source (recursively)
; Time complexity is O(n^2) in worst and average cases, O(n) - in best case
(defn f1 [n l]
  (if (pos? n)
    (let [i (rand-int (count l)), [xs x] (remove-at i l)]
      (cons x (f1 (dec n) xs)))
    '()))

File: haskell/p23.hs
-- P23 Extract a given number of randomly selected elements from a list.

import System.Random

-- Building a target list while removing items from the source (recursively)
-- Time complexity is O(n^2) in worst and average cases, O(n) - in best case
f1 :: RandomGen g => Int -> [a] -> g -> [a]
f1 0 _ _ = []
f1 n xs g = let (i, g') = randomR (0, length xs - 1) g
                (xs', x) = rm i xs
            in x : f1 (pred n) xs' g'

-- Helper function
rm :: Int -> [a] -> ([a], a)
rm n xs = let (a, x:b) = splitAt n xs in (a ++ b, x)